### PR TITLE
Render elevation bounds in maliput::utility::GenerateObj()

### PR DESCRIPTION
**Pull request by maddog-tri**
(This is the last elevation-bounds related PR for drake.)  (For now... bwah-ha-ha-ha.)

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/6263)

### CODE REVIEW

**Comment by maddog-tri** [comment]
+@liangfok for feature review, please.

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-:-Klz53ZbYoqiIj5FZxln:bltds52)*

**Comment by liangfok** [comment]
:lgtm: 

Using the default settings in view3dscene, I can barely see the bounds:



They become more discernible when I enable mesh visualization within view3dscene:



Perhaps in the future the bounds visualization can be made more discernable using the default settings and an actual volume surrounding the road as opposed to two surfaces.

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-:-Km4zY5I-DKpRArfE_q8:bc7flv3)*

**Comment by maddog-tri** [comment]
Everything is a lot more visible if you have a non-pitch-black background color. (File -> Preferences -> Default Background Color)



---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-:-Km7C6aqOpmemIA3CNza:b-iicyrj)*

**Comment by maddog-tri** [comment]
+@soonho-tri for platform review, please.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-:-Km7DY3icRIMPnQJqYvG:bxwr5c8)*

**Comment by soonho-tri** [comment]
Reviewed 6 of 8 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/utility/test/data/ChangeOrigin.obj, line 307 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-Km7rzAvTcrD2yEMQpsB:-Km7rzAvTcrD2yEMQpsC:blev0bw) ([raw file](https://github.com/robotlocomotion/drake/blob/6fa2a1dc2745f1e4289b81ab6c1165064b265759/drake/automotive/maliput/utility/test/data/ChangeOrigin.obj#L307)):*
> ```
> f 159//7 160//7 161//7 162//7
> f 162//5 160//5 163//5 164//5
> f 159//8 161//8 165//8 166//8
> ```

`*.obj` and `*.mtl` files in this directory are generated files and I think it's better to teach the fact to reviewable. Here are instructions: https://github.com/Reviewable/Reviewable/wiki/FAQ#how-do-i-tell-reviewable-that-a-file-is-generated-and-should-not-be-reviewed .

In short, we need `.gitattributes` file in this directory which includes:
```
*.obj -diff
*.mtl -diff
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6263)*

**Comment by soonho-tri** [comment]
Reviewed 2 of 8 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/utility/generate_obj.cc, line 300 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-Km7w_BBUf4D6kqtuWVN:-Km7w_BBUf4D6kqtuWVO:byhxslk) ([raw file](https://github.com/robotlocomotion/drake/blob/6fa2a1dc2745f1e4289b81ab6c1165064b265759/drake/automotive/maliput/utility/generate_obj.cc#L300)):*
> ```cpp
>                         double grid_unit,
>                         bool use_driveable_bounds,
>                         std::function<double(double, double)> elevation) {
> ```

BTW, please consider `const std::function<double(double, double)>& elevation`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6263)*

**Comment by soonho-tri** [comment]
:lgtm: (aside minor comments)

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-:-Km8EG993kk6vYwjWtqq:b-6v8yxj)*

**Comment by maddog-tri** [comment]
Comments addressed and/or replied to.  PTAL.

---

Review status: 7 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/utility/generate_obj.cc, line 300 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-Km7w_BBUf4D6kqtuWVN:-Km8QKQzK_AWEAhLKW50:b-f4erk8) ([raw file](https://github.com/robotlocomotion/drake/blob/6fa2a1dc2745f1e4289b81ab6c1165064b265759/drake/automotive/maliput/utility/generate_obj.cc#L300)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, please consider `const std::function<double(double, double)>& elevation`.
</blockquote></details>

Done.  (Though, it probably does very little here, since the callers are all passing lambdas anyway, which I imagine would only get wrapped in a `std::function` once whether passed by reference or by value.)

---

*[drake/automotive/maliput/utility/test/data/ChangeOrigin.obj, line 307 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-Km7rzAvTcrD2yEMQpsB:-Km8Qs5eLopaX268LaFy:b3qzd7n) ([raw file](https://github.com/robotlocomotion/drake/blob/6fa2a1dc2745f1e4289b81ab6c1165064b265759/drake/automotive/maliput/utility/test/data/ChangeOrigin.obj#L307)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

`*.obj` and `*.mtl` files in this directory are generated files and I think it's better to teach the fact to reviewable. Here are instructions: https://github.com/Reviewable/Reviewable/wiki/FAQ#how-do-i-tell-reviewable-that-a-file-is-generated-and-should-not-be-reviewed .

In short, we need `.gitattributes` file in this directory which includes:
```
*.obj -diff
*.mtl -diff
```
</blockquote></details>

Hmm... I'm not entirely comfortable with considering these to be "generated files".

The files in `automotive/gen` are definitely "generated":  there is a script which generates exactly those files.  Those generated files are compiled into other code, and tested either directly or via tests on code that uses the files.  The fact that those files are even checked into the source repository is evidence of a deficiency in the build system, which in a better world would be able to build them automagically without human intervention.

These `test/data` files here are different, because there is no automated way to generate them and, more importantly, because they are untestable --- they *are* the test.  They should not escape scrutiny during a review, and I worry that's what would happen if they were labeled "generated file".

A text-diff isn't always the best way to review them, true --- but it does contribute.  Here, the `.obj` files basically have a bunch of "h_bounds_haze" vertices/normals/faces appended to them.  And, the `.mtl` file clearly has one new stanza appended.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-:-Km8UZqSruPP3qTHk7jv:bt8z490)*

**Comment by soonho-tri** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/utility/test/data/ChangeOrigin.obj, line 307 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6263#-Km7rzAvTcrD2yEMQpsB:-Km8VI6hKjXLy9nUeTfh:bgfw5xp) ([raw file](https://github.com/robotlocomotion/drake/blob/6fa2a1dc2745f1e4289b81ab6c1165064b265759/drake/automotive/maliput/utility/test/data/ChangeOrigin.obj#L307)):*
> These test/data files here are different because there is no automated way to generate them and, more importantly, because they are untestable --- they are the test.

Thanks for the explanation. I was actually curious why we have both of the generator and the generated files. But now I understand that those `.obj` and `.mtl` files are the expected output to test the generator. Let's keep it as it is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6263)*
